bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs

```diff
@@ -179,6 +165,50 @@ select new
         return results;
     }
 
+    public (bool Read, bool Write) AccessToServiceAccounts(IEnumerable<Guid> ids, Guid userId,
+       AccessClientType accessType)
+    {
+        using var scope = ServiceScopeFactory.CreateScope();
+        var dbContext = GetDatabaseContext(scope);
+
+        var serviceAccount = dbContext.ServiceAccount.Where(sa => ids.Contains(sa.Id));
+        var query = CheckUserServiceAccountReadWriteAccess(accessType, userId, serviceAccount);
+        var policies = query.ToList();
+
+        if (!policies.Any())
+        {
+            return (false, false);
+        }
+
+        var read = policies.All(p => p.Read);
+        var write = policies.All(p => p.Write);
+
+        return (read, write);
+    }
+
+    private IQueryable<(bool Read, bool Write)> CheckUserServiceAccountReadWriteAccess(AccessClientType accessType, Guid userId, IQueryable<ServiceAccount> serviceAccount)
```

Comment:
```
  public async Task<(bool Read, bool Write)> AccessToServiceAccountsAsync(IEnumerable<Guid> ids, Guid userId, AccessClientType accessType)
    {
        using var scope = ServiceScopeFactory.CreateScope();
        var dbContext = GetDatabaseContext(scope);

        var serviceAccount = dbContext.ServiceAccount.Where(sa => ids.Contains(sa.Id));
        var query = ToAccessQuery(serviceAccount, userId, accessType);

        var policies = await query.ToListAsync();

        if (!policies.Any())
        {
            return (false, false);
        }

        var read = policies.All(p => p.Read);
        var write = policies.All(p => p.Write);

        return (read, write);
    }
    
    private class Access
    {
        public bool Read;
        public bool Write;
    }
    
    private static IQueryable<Access> ToAccessQuery(IQueryable<ServiceAccount> serviceAccount, Guid userId, AccessClientType accessType)
    {
        return accessType switch
        {
            AccessClientType.NoAccessCheck => serviceAccount.Select(_ => new Access { Read = true, Write = true }),
            AccessClientType.User => serviceAccount.Select(sa => new Access
            {
                Read = sa.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Read) ||
                       sa.GroupAccessPolicies.Any(ap =>
                           ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Read)),
                Write = sa.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Write) ||
                        sa.GroupAccessPolicies.Any(ap =>
                            ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Write)),
            }),
            AccessClientType.ServiceAccount => serviceAccount.Select(_ => new Access { Read = false, Write = false }),
            _ => serviceAccount.Select(_ => new Access { Read = false, Write = false }),
        };
    }
```

Was able to get something like this working. Instead of making a private class, we could also use `IQueryable<dynamic>` and cast the result in the public methods.

What do you think about the private class vs. dynamic?

Reply:
I'd personally prefer private class instead of dynamic